metaflow/plugins/airflow/airflow_utils.py

```diff
@@ -0,0 +1,410 @@
+
+
+LABEL_VALUE_REGEX = re.compile(r"^[a-zA-Z0-9]([a-zA-Z0-9\-\_\.]{0,61}[a-zA-Z0-9])?$")
+
+TASK_ID_XCOM_KEY = "metaflow_task_id"
+RUN_ID_LEN = 12
+TASK_ID_LEN = 8
+RUN_ID_PREFIX = "airflow"
+
+# AIRFLOW_TASK_ID_TEMPLATE_VALUE will work for linear/branched workflows.
+# ti.task_id is the stepname in metaflow code.
+# AIRFLOW_TASK_ID_TEMPLATE_VALUE uses a jinja filter called `task_id_creator` which helps
+# concatenate the string using a `/`. Since run-id will keep changing and stepname will be
+# the same task id will change. Since airflow doesn't encourage dynamic rewriting of dags
+# we can rename steps in a foreach with indexes (eg. `stepname-$index`) to create those steps.
+# Hence : Foreachs will require some special form of plumbing.
+# https://stackoverflow.com/questions/62962386/can-an-airflow-task-dynamically-generate-a-dag-at-runtime
+AIRFLOW_TASK_ID_TEMPLATE_VALUE = (
+    "%s-{{ [run_id, ti.task_id, dag_run.dag_id ] | task_id_creator  }}" % RUN_ID_PREFIX
+)
+
+
+class SensorNames:
+    EXTERNAL_TASK_SENSOR = "ExternalTaskSensor"
+    SQL_SENSOR = "SQLSensor"
+    S3_SENSOR = "S3KeySensor"
+
+    @classmethod
+    def get_supported_sensors(cls):
+        return list(cls.__dict__.values())
+
+
+def run_id_creator(val):
+    # join `[dag-id,run-id]` of airflow dag.
+    return hashlib.md5("-".join(val).encode("utf-8")).hexdigest()[:RUN_ID_LEN]
+
+
+def task_id_creator(lst):
+    # This is a filter which creates a hash of the run_id/step_name string.
+    # Since run_ids in airflow are constants, they don't create an issue with the
+    #
+    return hashlib.md5("/".join(lst).encode("utf-8")).hexdigest()[:TASK_ID_LEN]
+
+
+def json_dump(val):
+    return json.dumps(val)
+
+
+class AirflowDAGArgs(object):
+
+    # `_arg_types` is a dictionary which represents the types of the arguments of an Airflow `DAG`.
+    # `_arg_types` is used when parsing types back from the configuration json.
+    # It doesn't cover all the arguments but covers many of the important one which can come from the cli.
+    _arg_types = {
+        "dag_id": str,
+        "description": str,
+        "schedule_interval": str,
+        "start_date": datetime,
+        "catchup": bool,
+        "tags": list,
+        "dagrun_timeout": timedelta,
+        "default_args": {
+            "owner": str,
+            "depends_on_past": bool,
+            "email": list,
+            "email_on_failure": bool,
+            "email_on_retry": bool,
+            "retries": int,
+            "retry_delay": timedelta,
+            "queue": str,  #  which queue to target when running this job. Not all executors implement queue management, the CeleryExecutor does support targeting specific queues.
+            "pool": str,  # the slot pool this task should run in, slot pools are a way to limit concurrency for certain tasks
+            "priority_weight": int,
+            "wait_for_downstream": bool,
+            "sla": timedelta,
+            "execution_timeout": timedelta,
```

Comment:
What's the difference between `execution_timeout` and `dagrun_timeout`?

Reply:
Execution timeout is at task level and dagrun_timeout is at dag-run level. Hence `execution_timeout` is in `default_args`. `default_args` contain arguments which are defaults for all tasks in dag.

Reply:
Add comment on where we use defaults which are in parity with batch/stepfn/argo etc.